sirepo/simulation_db.py

```diff
@@ -963,15 +1012,62 @@ def _init():
         nfs_tries=(10, int, 'How many times to poll in hack_nfs_write_status'),
         nfs_sleep=(0.5, float, 'Seconds sleep per hack_nfs_write_status poll'),
     )
+    props = pkcollections.Dict()
+    props.source = 'external'
+    props.dir = 'ext'
+    LOADED_FILE_PROPS.append(props)
+    props = pkcollections.Dict()
+    props.source = 'local'
+    props.dir = ''
+    LOADED_FILE_PROPS.append(props)
+
 
+def _merge_dicts(base, derived, depth=1, merge_style=None):
+    """Copy the items in the base dictionary into the derived dictionary, to the specified depth
+    Items with the same name are not replaced
 
-def _merge_dicts(base, derived, depth=1):
-    if depth <= 0:
+    Args:
+        base (dict): source
+        derived (dict): receiver
+        depth (int): how deep to recurse:
+            >= 0:  <depth> levels
+            < 0:   all the way
+        merge_style (str): pass 'merge' to add elements of array in <base> to matching array in <derived>
+    """
+    if depth == 0:
         return
     for key in base:
         if key not in derived:
             derived[key] = base[key]
-        _merge_dicts(base[key], derived[key], depth - 1)
+        elif merge_style == 'merge':
+            try:
+                derived[key].extend(base[key])
+            except AttributeError:
+                # The value was not an array
+                pass
+        d = depth - 1 if depth > 0 else depth
+        try:
+            _merge_dicts(base[key], derived[key], d, merge_style)
+        except TypeError:
+            # The value in question is not itself a dict, move on
+            pass
+
+
+def _pkg_relative_path_static(file_dir, file_name):
+    """Path to a file under /static, relative to the package_data directory
+
+    Args:
+        file_dir (str): sub-directory of package_data/static
+        file_name (str): name of the file.  If starts with '/', just return file_name
+
+    Returns:
+        str: full relative path of the file
+    """
+    pkdlog('getting static path for {}:{}', file_dir, file_name)
+    if file_name[0] == '/':
```

Comment:
This was to handle the special case of calls to ```/user-state.js``` (actually a jinja template) that @moellep added as part of cookie management

Reply:
Perhaps this is an entirely different: externalModules, staticModules, dynamicModules. It's a very special case. That particular code looks very odd, especially since there are paths being generated.

Reply:
Here's another [DesignHint](https://github.com/radiasoft/pykern/wiki/DesignHints#learn-libraries-theres-an-api-for-that) which discusses the problem. I think static_file_path should be privatized and made _static_file_uri.